catalyst-gateway/tests/api_tests/utils/rbac_chain.py

```diff
@@ -0,0 +1,102 @@
+from datetime import datetime, timezone
+import base64
+import pytest
+from enum import IntEnum
+import json
+from pycardano.crypto.bip32 import BIP32ED25519PrivateKey, BIP32ED25519PublicKey
+
+with open("./test_data/rbac_regs/only_role_0.jsonc", "r") as f:
+    ONLY_ROLE_0_REG_JSON = json.load(f)
+with open("./test_data/rbac_regs/role_3.jsonc", "r") as f:
+    ROLE_3_REG_JSON = json.load(f)
+
+
+class RoleID(IntEnum):
+    ROLE_0 = 0
+    PROPOSER = 3
+
+    def __str__(self):
+        return f"{int(self)}"
+
+
+class RBACChain:
+    def __init__(self, keys_map: dict, network: str):
+        # corresponded to different roles bip32 extended ed25519 keys map
+        self.keys_map = keys_map
+        self.network = network
+
+    def auth_token(self) -> str:
+        role_0_keys = self.keys_map[f"{RoleID.ROLE_0}"]
+        return generate_rbac_auth_token(
+            "cardano", self.network, role_0_keys["pk"], role_0_keys["sk"]
+        )
+
+    # returns a role's catalyst id, with the provided role secret key
+    def cat_id_for_role(self, role_id: RoleID) -> (str, str):
+        role_data = self.keys_map[f"{role_id}"]
+        role_0_pk = self.keys_map[f"{RoleID.ROLE_0}"]["pk"]
+        return (
+            generate_cat_id(
+                "cardano",
+                self.network,
+                role_id,
+                role_0_pk,
+                role_data["rotation"],
+            ),
+            role_data["sk"],
+        )
+
+
+@pytest.fixture
+def rbac_chain_factory():
+    def __rbac_chain_factory(role_id: RoleID) -> RBACChain:
+        network = "preprod"
+        match role_id:
+            case RoleID.ROLE_0:
+                return RBACChain(ONLY_ROLE_0_REG_JSON, network)
+            case RoleID.PROPOSER:
+                return RBACChain(ROLE_3_REG_JSON, network)
+
+    return __rbac_chain_factory
+
+
+def generate_cat_id(
+    network: str, subnet: str, role_id: RoleID, pk_hex: str, rotation: int
+):
+    pk = bytes.fromhex(pk_hex)[:32]
+    prefix = "catid.:"
+    nonce = int(datetime.now(timezone.utc).timestamp())
+    subnet = f"{subnet}." if subnet else ""
+    role0_pk_b64 = base64_url(pk)
+
+    if role_id == RoleID.ROLE_0:
+        return f"{prefix}{nonce}@{subnet}{network}/{role0_pk_b64}"
+
+    return f"{prefix}{nonce}@{subnet}{network}/{role0_pk_b64}/{role_id}/{rotation}"
+
+
+def generate_rbac_auth_token(
+    network: str,
+    subnet: str,
+    pk_hex: str,
+    sk_hex: str,
+) -> str:
+    pk = bytes.fromhex(pk_hex)[:32]
+    sk = bytes.fromhex(sk_hex)[:64]
+    chain_code = bytes.fromhex(sk_hex)[64:]
+
+    bip32_ed25519_sk = BIP32ED25519PrivateKey(sk, chain_code)
+    bip32_ed25519_pk = BIP32ED25519PublicKey(pk, chain_code)
+
+    cat_id = generate_cat_id(network, subnet, RoleID.ROLE_0, pk_hex, 0)
```

Comment:
The role and rotation should be omitted for auth token
> role and rotation are not specified (defaults are always used, and not encoded)
Ref: https://github.com/input-output-hk/catalyst-voices/blob/main/docs/src/catalyst-standards/permissionless-auth/auth-header.md#backend-processing-of-the-token

The auth token validation is wrong, fixed it in my current task - Rbac auth token integration test

Reply:
Yes you are right inside the `generate_cat_id` function we have such condition
```
if role_id == RoleID.ROLE_0 and rotation == 0:
 return f"{prefix}{nonce}@{subnet}{network}/{role0_pk_b64}"
```